util/server/index.ts

```diff
@@ -0,0 +1,149 @@
+    headers: { 'Accept-Type': 'text/plain' }
+  });
+  const devfileYAML = (await res.text()) as string;
+  return devfileYAML;
+};
+
+const getLocalYAML = async (devfileName: string, yamlLocation: string): Promise<string> => {
+  const devfileYAMLRelPath = `${yamlLocation}/${devfileName}/devfile.yaml`.split('/');
+  const devfileYAMLAbsPath = path.join(process.cwd(), ...devfileYAMLRelPath);
+  const devfileYAML = (await fs.readFile(devfileYAMLAbsPath, 'utf8')) as string;
+  return devfileYAML;
+};
+
+const getENVHosts = () => {
+  const envHosts = process.env.DEVFILE_REGISTRY_HOSTS?.split('|').filter((host) => host !== '');
+
+  let hosts: Host = {};
+
+  if (envHosts?.length) {
+    envHosts.forEach((envHost) => {
+      const [hostName, sourceType, hostLocation] = envHost.split('>');
+
+      if (!(sourceType === 'url' || sourceType === 'stacks')) {
+        throw Error(
+          'The environment variable DEVFILE_REGISTRY_HOSTS can only accept "url" or "stacks"'
+        );
+      }
+
+      hosts = {
+        ...hosts,
+        [hostName]: {
+          [sourceType]: hostLocation
+        }
+      };
+    });
+  }
+
+  return hosts;
+};
+
+export const getDevfilesJSON = async (): Promise<Devfile[]> => {
+  let hosts: Host = await getConfigFileHosts('/config/devfile-registry-hosts.json');
+  hosts = { ...hosts, ...getENVHosts() };
+
+  if (!Object.keys(hosts).length) {
+    hosts = {
+      'devfile.io': {
+        url: 'https://registry.devfile.io'
+      }
+    };
+  }
+
+  let devfiles: Devfile[] = [];
+  await Promise.all(
+    Object.entries(hosts).map(async ([hostName, hostLocation]) => {
+      let extractedDevfiles: Devfile[] = [];
+      if (hostLocation.url) {
+        extractedDevfiles = await getRemoteJSON(hostName, hostLocation.url);
+      }
+      if (hostLocation.stacks) {
+        extractedDevfiles = await getLocalJSON(hostName, hostLocation.stacks);
+      }
+
+      if (extractedDevfiles.length) {
+        devfiles = devfiles.concat(extractedDevfiles);
+      }
+    })
+  );
+
+  return devfiles;
+};
+
+export const getDevfileYAML = async (devfile: Devfile): Promise<getDevfileYAMLReturnType> => {
+  let devfileYAML = null;
+  let devfileJSON = null;
+
+  if (devfile.type !== 'stack') {
+    return { devfileYAML, devfileJSON };
+  }
+
+  let hosts: Host = await getConfigFileHosts('/config/devfile-registry-hosts.json');
+  hosts = { ...hosts, ...getENVHosts() };
+
+  if (!Object.keys(hosts).length) {
+    hosts = {
+      'devfile.io': {
+        url: 'https://registry.devfile.io'
+      }
+    };
+  }
+
+  for (const [hostName, hostLocation] of Object.entries(hosts)) {
+    if (hostName === devfile.sourceRepo) {
+      if (hostLocation.url) {
+        devfileYAML = await getRemoteYAML(devfile.name, hostLocation.url);
+      }
+      if (hostLocation.stacks) {
+        devfileYAML = await getLocalYAML(devfile.name, hostLocation.stacks);
+      }
+    }
+  }
```

Comment:
this logic is almost the same as the other func. Can they be abstracted? Maybe call the other func? And they can have a flag whether to parse as JSON/YAML

Reply:
Yeah I can look into trying to abstract it.

Reply:
I dont think this is worth while after adding type checking